src/zend/mod.rs

```diff
@@ -62,6 +63,81 @@ pub fn printf(message: &str) -> Result<()> {
     Ok(())
 }
 
+/// Writes binary data to PHP's output stream (stdout).
+///
+/// Unlike [`printf`], this function is binary-safe and can handle data
+/// containing NUL bytes. It uses the SAPI module's `ub_write` function
+/// which accepts a pointer and length, allowing arbitrary binary data.
+///
+/// Also see the [`php_write!`] macro.
+///
+/// # Arguments
+///
+/// * `data` - The binary data to write to stdout.
+///
+/// # Returns
+///
+/// The number of bytes written, or 0 if the SAPI's `ub_write` function
+/// is not available.
+///
+/// # Example
+///
+/// ```no_run
+/// use ext_php_rs::zend::write;
+///
+/// // Write binary data including NUL bytes
+/// let data = b"Hello\x00World";
+/// write(data);
+/// ```
+#[must_use]
+pub fn write(data: &[u8]) -> usize {
+    unsafe {
+        if let Some(ub_write) = sapi_module.ub_write {
+            ub_write(data.as_ptr().cast::<c_char>(), data.len())
+        } else {
+            0
+        }
+    }
+}
+
+/// Writes binary data to stderr.
+///
+/// This function is binary-safe and can handle data containing NUL bytes.
+/// It writes directly to stderr, bypassing PHP's error handling which
+/// uses null-terminated strings.
+///
+/// Also see the [`php_write_err!`] macro.
+///
+/// # Arguments
+///
+/// * `data` - The binary data to write to stderr.
+///
+/// # Returns
+///
+/// The number of bytes written, or 0 if writing failed.
+///
+/// # Example
+///
+/// ```no_run
+/// use ext_php_rs::zend::write_err;
+///
+/// // Write binary data including NUL bytes to stderr
+/// let data = b"Error: \x00binary\x00data";
+/// write_err(data);
+/// ```
+#[must_use]
+pub fn write_err(data: &[u8]) -> usize {
+    use std::io::Write;
+
+    if data.is_empty() {
+        return 0;
+    }
+
+    std::io::stderr()
+        .write(data)
+        .unwrap_or(0)
+}
```

Comment:
I don't see any benefit to having this wrapper in our lib.
Maybe a small doc is enough